Ad5770r iio driver

mirceacaprioru opened the pull request.

commodo commented on MAINTAINERS:
```diff
@@ -790,6 +790,15 @@ S:	Supported
+F:	drivers/iio/dac/ad5770r.c
+F:	Documentation/devicetree/bindings/iio/dac/ad5770r.txt
+
+
```
extra line

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+		return ad5770r_soft_reset(st);
+	}
+
+	mdelay(1);
```
upstream people would say: "Explanation must be given for such a long delay without idling." :)

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+
+	st->gpio_reset = devm_gpiod_get_optional(&st->spi->dev, "reset",
+						 GPIOD_OUT_HIGH);
+
```
extra line

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,101 @@
+		* Byte 2 - IB_EXT_EN
+		* Byte 3 - MON_CH
+
+- adi,ref_voltage: Reference voltage selection.
```
do not use '_'; also properties which have a unit if measure are recommended to have unit suffix appended to the property name: i.e: ref-voltage-microvolt : https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/property-units.txt

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+		const struct ad5770r_channel_switches *channel_enable)
+{
+	if (!st || !channel_enable)
+		return ENOMEM;
```
typically errors are returned as negative values ; so  `return -ENOMEM` ; this practice is used is to differentiate between positive values that mean "data read" or "states", 0 [generally means success] and negative [ generally error codes ]

in this case, it should be  `-EINVAL` ; since it means invalid argument/value

commodo replied:
though, in this case, since the function is static[private], you could skip the error checking, since you could be sure [ from the code path] that `st` and `channel_enable` are non-NULL;

this comment is available for a few other places;

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+
+	if (!device_property_read_u8_array(&st->spi->dev,
+		"adi,channel_cfg", tmparray, 7)) {
+		st->ch_config.en0 = tmparray[0];
```
why not make `en0,en1,...en5` an array ?
it could save some lines by iterating over the `en` array in some places;

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+
+	ad5770r_update_range(st);
+
+	if (!device_property_read_u8_array(&st->spi->dev, "adi,alarm",
```
general comment/question for these device-tree properties: do we need to throw an error for some of them ? or can all of the be optional ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+		ad5770r_min_max_table[AD5770R_CH2].max = 150;
+	}
+
+	for (i = AD5770R_CH2; i < AD5770R_MAX_CHANNELS; i++)
```
i'd put brackets to this for loop;

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+{
+	int i;
+
+	if (st->output_mode[AD5770R_CH0].output_range_mode == 0x00) {
```
could the values for `output_range_mode` be made an enum to make it clearer what they mean ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+	u8 buf[2];
+
+	switch (info) {
+	case IIO_CHAN_INFO_RAW:
```
I'm curios here; do we need locking here and similar places ?
I'm a bit vague on this matter; maybe @lclausen-adi could give an idea.

lclausen-adi replied:
I don't think this needs locking. regmap has internal locking so the state should always be consistent. When issuing multiple but related regmap_read()/regmap_write() calls locking is required to ensure that the state is consistent.

commodo replied:
thanks for the input :)

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+	if (!device_property_read_u8_array(&st->spi->dev, "adi,alarm",
+					    tmparray, 8)) {
+		st->alarm_config.open_drain_en = tmparray[0];
+		st->alarm_config.thermal_shutdown_en = tmparray[1];
```
for these variables and similar, we probably need a different approach;
it's not very nice if we read an array and pass entries from the array into the struct members;

i'm a bit unsure about which other approach to use atm;

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,715 @@
+		st->ch_config.en3 = tmparray[3];
+		st->ch_config.en4 = tmparray[4];
+		st->ch_config.en5 = tmparray[5];
+	}
```
All of these properties are optional, but what happens if they are not defined in the devicetree? There is no default value for them.

commodo commented:
for visibility: i've update the `rpi-4.14.y` base with updates from upstream: Raspberry Pi, Xilinx [Rebase 4.14] and ADI master [that were added recently]

the reason for doing this is also because I had to force-update the rpi-4.14.y anyway;
commit eee97d12665fef8cf429a1e5035b23ae969705b8 breaks some drivers, and we need to see whether we revert it [for 4.14] or update it over all drivers

mirceacaprioru commented:
Changelog V1->V2

- replace struct en0 en5 with bool array
- added checks for dts properties
- commented delays in reset function
- added microvolt unit to dts and documentation
- used unsigned char for consistency

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+
+/* AD5770R configuration registers */
+#define AD5770R_CHANNEL_CONFIG		0x14
+#define AD5770R_OUTPUT_RANGE_CH0	0x15
```
you could compress this into a parameterize-able macro like `AD5770R_OUTPUT_RANGE(ch) (0x15 + ch)`  and compute them;

it looks like you aren't using all definitions here;
you could potentially remove a few

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+#define AD5770R_OUTPUT_RANGE_CH5	0x1A
+#define AD5770R_REFERENCE		0x1B
+#define AD5770R_ALARM_CONFIG		0x1C
+#define AD5770R_OUTPUT_FILTER_CH0	0x1D
```
similarly here; `AD5770R_OUTPUT_FILTER(ch)   (0x1D + (ch))`

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+#define AD5770R_MONITOR_SETUP		0x23
+#define AD5770R_STATUS			0x24
+#define AD5770R_HW_LDAC			0x25
+#define AD5770R_CH0_DAC_LSB		0x26
```
these as well;  `AD5770R_DAC_LSB(ch)  (0x26 + (ch))`  and same for _MSB

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+};
+
+struct ad5770r_output_modes {
+
```
nitpick: extra line

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+	AD5770R_TEMPERATURE_MONITORING
+};
+
+enum ad5770r_ch {
```
looking at this a bit, it feels like this enum is redundant; 
since you have 6 channels, you can just use `int channel` and do some basic checks of whether `(ch >= 0 && ch < AD5770R_MAX_CHANNELS)` where needed

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+					      &st->output_mode[i], i);
+
+	ret = ad5770r_set_alarm(st, &st->alarm_config);
+
```
nitpick: you could remove empty lines in constructs like this:
```
ret = func();
<line>
if (ret < 0)
    return ret;
```

the empty line does not help much with readability in this case;

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+
+static int ad5770r_check_output_mode(struct ad5770r_state *st)
+{
+	int ret, i;
```
`ret` is uninitialized here

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+
+		ret = ad5770r_check_output_mode(st);
+
+		if (ret)
```
shouldn't it be the other way around ?
`if (ret == 0)` then read values ?
you could also use `memcpy()` here

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+		return IIO_VAL_INT;
+	case IIO_CHAN_INFO_SCALE:
+		ret = ad5770r_search_mode(st, chan->channel, &min, &max);
+
```
you could remove this line

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,101 @@
+                adi,ref-voltage-microvolt = [01];
+                adi,external-ref = [00];
+                adi,channel-cfg = [01 01 01 01 01 01 00];
+                adi,channel-enable = [01 01 01 01 01 01];
```
You should not have a device property that ends in a verb. I recently learned this myself, thanks to @larsclausen :)
Instead, you can use iio_chan_spec_ext_info struct to enable or disable your DAC channels. Here's an example:
https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y/drivers/iio/dac/ad5686.c#L186

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+{
+	int i;
+
+	for (i = 0; i < AD5770R_MAX_CH_MODES; i++)
```
maybe add brackets to this for loop ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+{
+	unsigned char regval;
+
+	regval = AD5770R_ITF_CFG_A_SW_RESET(1);
```
if you're not planning to add anything here, you could remove the extra variable and wrap all this in 1-2 lines;

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,101 @@
+		These are the valid values:
+			00 - internal resistor
+			01 - external resistor
+
```
The typical way to define an external vref is by using the [fixedregulator](https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/regulator/fixed-regulator.txt) property.
You can define the internal ref inside a chip_info struct and use it only if the vref property is missing from the DT. You can follow this example: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad5380.c#L404

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,101 @@
+		Available options:
+			00 - output shutdown (disable for sink)
+			01 - normal operation (enable for sink)
+
```
You can configure the channels during channel enable. This looks redundant to me.

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+		buf[0] = ((u16)val >> 6) & 0xFF;
+		buf[1] = (val & 0x3F) << 2;
+		ret = regmap_bulk_write(st->regmap, chan->address,
+					buf, 2);
```
You can use ARRAY_SIZE here and return directly, therefore there is no need for the ret variable:
`return regmap_bulk_write(st->regmap, chan->address, buf, ARRAY_SIZE(buf));`

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,716 @@
+		return -EINVAL;
+	}
+
+	return ret;
```
By following the comment above, this return will never be reached.

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,76 @@
+		* Channel 3 to Channel 5:
+			* 00: 0 to 45 mA
+			* 01: 0 to 100 mA
+
```
I think we should define this property in a different way: 
ad5770r node should contain child nodes representing each channel. Inside the child node, we should include a property that defines the DAC range in microamps. Something like this:
```
	channel@0 {
		adi,range-output-microamp = <0 300000>
	};
```
Here is an example: https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/iio/dac/ad5592r.txt#L84

stefpopa commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,76 @@
+		* Byte 1 - MUX_BUFFER
+		* Byte 2 - IB_EXT_EN
+		* Byte 3 - MON_CH
+
```
I think that adi,alarm and adi,monitor properties should not be part of the DT because they do not describe the hardware.

lclausen-adi replied:
:thumbsup:

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,672 @@
+	bool readin;
+	int ret;
+
+	ret = strtobool(buf, &readin);
```
please use kstrtobool

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,672 @@
+		buf[1] = (val & 0x3F) << 2;
+		return regmap_bulk_write(st->regmap, chan->address,
+					 buf, ARRAY_SIZE(buf));
+		break;
```
some compilers [or compiler options] may complain about this being an "unreachable statement";
you can remove this break statement;

commodo commented on arch/arm/boot/dts/overlays/rpi-ad5770r-overlay.dts:
```diff
@@ -0,0 +1,44 @@
+	};
+};
+
+
```
nitpick: you can remove all extra new lines

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,672 @@
+#define AD5770R_ITF_CFG_A_ADDR_ASC_MSB(x)	(((x) & 0x1) << 5)
+
+/* AD5770R_INTERFACE_CONFIG_B */
+#define AD5770R_ITF_CFG_B_SINGLE_INST(x)	(((x) & 0x1) << 7)
```
this [and maybe other macros] look(s) unused
will you use them later ?

mirceacaprioru commented:
Changelog V2->V4
- added channel child nodes to dts
- removed alarm and monitor device tree bindings

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+	struct ad5770r_out_range	output_mode[AD5770R_MAX_CHANNELS];
+	int				vref;
+	bool				internal_ref;
+	bool				ch_pwr_down[6];
```
why not use `AD5770R_MAX_CHANNELS` here instead of 6 ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+	int ret, tmp[2], min, max, i;
+	unsigned int index;
+	struct fwnode_handle *child;
+	bool ch_config[6] = {0};
```
AD5770R_MAX_CHANNELS here ? instead of 6 ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+	for (i = 0; i < AD5770R_MAX_CHANNELS; i++)
+		ret = ad5770r_set_output_mode(st,
+					      &st->output_mode[i], i);
+	if (ret)
```
this only checks if the last channel errored;
maybe move this into the loop ?

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+	if (ret)
+		return ret;
+
+	memset(st->ch_pwr_down, 1, ARRAY_SIZE(st->ch_pwr_down));
```
this looks a bit funky ; 
are all `ch_pwr_down` supposed to be inited to true ?
I'm not sure if a simple for loop is better here; since `ch_pwr_down` is of type `bool`

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+
+	st->output_mode[index].out_range_mode = mode;
+
+	return ret;
```
this whole function could be simplified a bit; something like

```
for (i = 0; i < AD5770R_MAX_CH_MODES; i++) {
      if (ad5770r_rng_tbl[i].ch != index)
          continue;
	if (ad5770r_rng_tbl[i].min != min ||
            ad5770r_rng_tbl[i].max != max)
            continue;
       st->output_mode[i].out_range_mode = ad5770r_rng_tbl[i].mode;
       return 0;
}

return -EINVAL;
```

also, maybe it should be renamed to `ad5770r_store_output_range` or `ad5770r_validate_output_range` ;
this doesn't set the output range in HW, but rather stores it on the state struct and does a validation first

commodo commented on arch/arm/boot/dts/overlays/rpi-ad5770r-overlay.dts:
```diff
@@ -0,0 +1,67 @@
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+vref:
```
nitpick: this should be indented ; you could put it on the same line with `fixedregulator@0`

commodo commented on arch/arm/boot/dts/overlays/rpi-ad5770r-overlay.dts:
```diff
@@ -0,0 +1,67 @@
+	fragment@0 {
+		target = <&spi0>;
+		__overlay__ {
+#address-cells = <1>;
```
nitpick: indentation

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,85 @@
+	- compatible: Must be "adi,ad5770r"
+	- reg: SPI chip select number for the device
+	- spi-max-frequency: Max SPI frequency to use (< 10000000)
+	-child nodees: Each child node represents one channel and has
```
nitpick: typo ; and add extra space `- child nodes`

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,549 @@
+
+	for (i = 0; i < AD5770R_MAX_CHANNELS; i++)
+		if (!ch_config[i])
+			return -EINVAL;
```
why not allow certain channels to not be used ?
if the user wants to use only 2 ?

mirceacaprioru replied:
the user has the option to use only  2 channels since they are in powerdone initialy. For safety purposes the channels configurations is no longer optional and the user must specify a mode for all channels.

commodo replied:
sounds good :)
thanks for the info

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,85 @@
+		reset-gpios = <&gpio 22 0>;
+
+		channel@0 {
+			index = <0>;
```
[more as a note] this is fine as is; i.e. the `index` keyword;
you could use the `reg` keyword/construct for this purpose;
it seems more like a convention, rather than a rule ;
but there are functions to support this when writing the driver this way;
see: https://elinux.org/Device_Tree_Usage#How_Addressing_Works

or, you could rename `index` to `num` ; it sounds better than `index` in the device-tree

mirceacaprioru commented:
Changelog V4 -> V5
- changed function set_output_mode to store_output_mode
- added return checks to init function inside for loop
- initialized ret variable to 0
- added min range value to 0 since the current sinking part is not controlled by the DAC

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,545 @@
+			break;
+		default:
+			regval = AD5770R_REF_RESISTOR_SEL(st->internal_ref) |
+				 AD5770R_REF_SEL(AD5770R_INT_1_25_V_OUT_OFF);
```
for most cases this is fine ;
i'd add a break statement, so that some compiler options won't complain here

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,545 @@
+module_spi_driver(ad5770r_driver);
+
+MODULE_AUTHOR("Mircea Caprioru <mircea.caprioru@analog.com>");
+MODULE_DESCRIPTION("Analog Devices AD5770 IDAC");
```
nitpick: `AD5770` -> `AD5770R` ?
since this is the module name it may make sense to keep things consistent

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -0,0 +1,545 @@
+	bool ch_config[AD5770R_MAX_CHANNELS] = {0};
+
+	device_for_each_child_node(&st->spi->dev, child) {
+		ret = fwnode_property_read_u32(child, "index", &index);
```
you should change this to "num" if you've updated the device-trees

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5770r.txt:
```diff
@@ -0,0 +1,85 @@
+Required properties for the AD5770R:
+	- compatible: Must be "adi,ad5770r"
+	- reg: SPI chip select number for the device
+	- spi-max-frequency: Max SPI frequency to use (< 10000000)
```
you could reference the SPI docs here ; SPI drivers seem to do this:
so, smth like:
```
        - spi-max-frequency: Max SPI frequency to use (< 10000000, as per
                                          Documentation/devicetree/bindings/spi/spi-bus.txt
```

commodo commented:
rest looks good :)

mirceacaprioru commented:
V5->V6
- replaced index variable with enum
- solved sparse warnings
